modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java

```diff
@@ -113,11 +133,23 @@ public class GeoIpDownloader extends AllocatedPersistentTask {
         this.client = client;
         this.clusterService = clusterService;
         this.threadPool = threadPool;
+        this.eagerDownload = EAGER_DOWNLOAD_SETTING.get(settings);
+        this.atLeastOneGeoipProcessor = hasAtLeastOneGeoipProcessor(clusterService.state());
         endpoint = ENDPOINT_SETTING.get(settings);
         pollInterval = POLL_INTERVAL_SETTING.get(settings);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(EAGER_DOWNLOAD_SETTING, this::setEagerDownload);
         clusterService.getClusterSettings().addSettingsUpdateConsumer(POLL_INTERVAL_SETTING, this::setPollInterval);
```

Comment:
We talked about this outside of the PR messaging, but we both think this could lead to some messy leaks.

Every time this task is allocated to a node, these settings listeners are being installed. There is no way to remove them and it doesn't look like they correctly check if the task is cancelled before rescheduling the background logic when they change. We should definitely protect against multiple downloaders running when updating either of these settings.

We could add a check to the update methods to make sure the task is active, but this still has the possibility of leaving a number of leaked task instances in the cluster service on any nodes that have previously cancelled them. We agreed that it would be better if we added a settings container of some sort that keeps an instance to the active downloader task on a node and calls it during updates if it exists. The settings container is light weight and can stick around without too much overhead since there will only ever be one of them, but the downloaders can come and go without leaking multiple update consumers.

Reply:
The summary of another offline conversation:
It's just a bad idea to have these relatively-short-lived objects like GeoIpDownloader be settings or cluster state listeners because:

1. You have to be careful to un-listen when you're done so that you don't wind up with leaks
2. You have increased risk of race conditions since the condition you are listening for could happen in between the time the object is created and the time it is registered as a listener (not to mention when its predecessor is un-registered).

So I've moved all listeners for settings and cluster state changes into the executor. The task gets the current values for these things from the executor when it needs them, and the executor handles telling the task when it needs to reschedule because of a change in a dynamic property.